gapic/src/main/com/google/gapid/perfetto/views/TrackContainer.java

```diff
@@ -52,20 +52,40 @@
   private TrackContainer() {
   }
 
+  public static enum Truncate {
+    left {
+      @Override
+      public void drawTextTruncate(RenderContext ctx, Fonts.Style style, String text,
+          double x, double y, double w, double h) {
+        ctx.drawTextLeftTruncate(Fonts.Style.Normal, text, x, y, w, h);
+      }
+    },
+    right {
+      @Override
+      public void drawTextTruncate(RenderContext ctx, Fonts.Style style, String text,
+          double x, double y, double w, double h) {
+        ctx.drawTextRightTruncate(Fonts.Style.Normal, text, x, y, w, h);
+      }
+    };
+
+    public abstract void drawTextTruncate(RenderContext ctx, Fonts.Style style, String text,
+        double x, double y, double w, double h);
+  }
+
   public static <T extends TrackPanel<T>> TrackConfig.Track.UiFactory<Panel> single(
-      TrackConfig.Track.UiFactory<T> track, boolean sep) {
-    return state -> new Single<T>(state, track.createPanel(state), sep, null, true);
+      TrackConfig.Track.UiFactory<T> track, boolean sep, Truncate truncate) {
```

Comment:
Would be simpler if this is just a boolean.  You can also overload the 'single' function to default to truncate left.

Reply:
Yes but it will look very hacky. This would be a cleaner approach and we can add more if we want.

Reply:
I think this is more readable

Reply:
Old but gold: https://ariya.io/2011/08/hall-of-api-shame-boolean-trap

Reply:
Yes, unfortunately, Java does not support named argument.  But that is an easily solvable problem.  We can just add a comment so that it's look like `single(..., true /*truncateLeft*/)`.  Or pass a constant enum value so it looks like `single(..., Truncate.LEFT)`.  Even the page you link suggest different ways to named the argument instead of turning the boolean argument into a functor.

The PR unnecessarily complicate the code for flexibility that is not currently needed.  You have to go through multiple object inheritance to determine what truncate.drawText actually do.  I mean...if you think about it, it's not even named well.  The functor class is named Truncate, but the actual main purpose is drawing text.

Not worth arguing over it if that doesn't convince you, but I do think it's make the code more complicated than it needs to be.

Reply:
maybe just change the name to `DrawTextTruncate`

Reply:
I had it first as a simple enum. My idea to move this to a enum method was because the caller is actually a long line of code. It will look a little ugly if we have multiple lines looking exactly same except for the change of the function name. Another argument was that, if we want to do a different justification, we can just add a new LeftJustified or RightJustified in the enum and call whatever function needs to be called there. I can change it to just a simple enum with if else statement if its needed. I believe this just looks cleaner.

Reply:
Would also prefer it to go with a simple boolean value + if else statement at rendering. 
1.The code style of `single()` track creation method. The other lightweight variable for initialization are in boolean, e.g. filtered or not, expanded or not, with separator or not.
2.The code style of AGI, or JAVA. So far the enum definition only happens when there's more than two possibilities for a state. Since it's either left or right, the creation of a new class carrying function inside is over-killing.

Reply:
> Another argument was that, if we want to do a different justification, we can just add a new LeftJustified or RightJustified in the enum and call whatever function needs to be called there.

That's what I mean by complicating code for flexibility that is not currently needed.  And frankly, using a simple enum does the same thing.  If you use simple enum, adding new draw function is just as easy.  You would just add a new case in a switch statement.

The typical reason to use a functor is to allow the caller to implement it.  But this is an enum, so caller cannot implement their own draw function.  Nor do we actually to give this flexibility to the caller.

> I'll change it to just a simple enum with if else statement if you guys think so. I believe this just looks cleaner.

I have to disagree, but not strongly enough that I will block it if you insist.

Reply:
I've removed the enum method and made it simple if-else inside RenderContext now. But kept the enum for defining left or right.